pkg/distributor/query.go

```diff
@@ -264,59 +311,107 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 				return nil, err
 			}
 
-			// Enforce the max chunks limits.
-			if chunkLimitErr := queryLimiter.AddChunks(resp.ChunksCount()); chunkLimitErr != nil {
-				return nil, validation.LimitError(chunkLimitErr.Error())
-			}
+			if len(resp.Timeseries) > 0 || len(resp.Chunkseries) > 0 {
+				// Enforce the max chunks limits.
+				// TODO: enforce chunk limit for streaming series
+				if chunkLimitErr := queryLimiter.AddChunks(ingester_client.ChunksCount(resp.Chunkseries)); chunkLimitErr != nil {
+					return nil, validation.LimitError(chunkLimitErr.Error())
+				}
 
-			for _, series := range resp.Chunkseries {
-				if limitErr := queryLimiter.AddSeries(series.Labels); limitErr != nil {
-					return nil, validation.LimitError(limitErr.Error())
+				for _, series := range resp.Chunkseries {
+					if limitErr := queryLimiter.AddSeries(series.Labels); limitErr != nil {
+						return nil, validation.LimitError(limitErr.Error())
+					}
 				}
-			}
 
-			if chunkBytesLimitErr := queryLimiter.AddChunkBytes(resp.ChunksSize()); chunkBytesLimitErr != nil {
-				return nil, validation.LimitError(chunkBytesLimitErr.Error())
-			}
+				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(ingester_client.ChunksSize(resp.Chunkseries)); chunkBytesLimitErr != nil {
+					return nil, validation.LimitError(chunkBytesLimitErr.Error())
+				}
 
-			for _, series := range resp.Timeseries {
-				if limitErr := queryLimiter.AddSeries(series.Labels); limitErr != nil {
-					return nil, validation.LimitError(limitErr.Error())
+				for _, series := range resp.Timeseries {
+					if limitErr := queryLimiter.AddSeries(series.Labels); limitErr != nil {
+						return nil, validation.LimitError(limitErr.Error())
+					}
 				}
-			}
 
-			// This goroutine could be left running after replicationSet.Do() returns,
-			// so check before writing to the results chan.
-			select {
-			case <-stop:
+				// This goroutine could be left running after replicationSet.Do() returns,
+				// so check before writing to the results chan.
+				select {
+				case <-stop:
+					return nil, nil
+				case results <- resp:
+				}
+			} else if len(resp.Series) > 0 {
+				seriesLabels := make([]labels.Labels, 0, len(resp.Series))
+
+				for {
+					for _, s := range resp.Series {
+						if limitErr := queryLimiter.AddSeries(s.Labels); limitErr != nil {
+							return nil, validation.LimitError(limitErr.Error())
+						}
+
+						seriesLabels = append(seriesLabels, mimirpb.FromLabelAdaptersToLabels(s.Labels))
+					}
+
+					if resp.IsEndOfSeriesStream {
+						break
+					}
+
+					resp, err = stream.Recv()
```

Comment:
Design wise, it looks weird to me having a for loop here which iterates over `stream.Recv()`, considering there's already an outer loop iterating over `stream.Recv()`. I think we can do some refactoring here.

Reply:
Yep, I want to revisit this whole method, it's become enormous and I suspect can be refactored.

Reply:
I've refactored this in 70e788c - there's still some room for improvement and I'll come back to it later, but I've simplified the cleanup logic and removed this nested loop as well.